### PR TITLE
Feature/extra fields

### PR DESCRIPTION
**Pull request by geovations**

### CODE REVIEW

File: cli/src/CliProcessor.cpp [hunk 1]
```diff
@@ -713,6 +714,10 @@ void CliProcessor::readOutputArgs(variables_map vm, bool splitArgs)
     osnArgs.dgcsCatalogID = vm.find("dgcs-catalog-id") != end(vm);
     osnArgs.evwhsCatalogID = vm.find("evwhs-catalog-id") != end(vm);
     readVariable("wfs-credentials", vm, osnArgs.wfsCredentials);
+    readVariable("extra-fields", vm, osnArgs.extraFields, true);
+    if (osnArgs.extraFields.size() > 0 && osnArgs.extraFields.size() % 2 != 0){
```
Comment (rddesmond) [inline cli/src/CliProcessor.cpp]: Don't need the first check (`0 % 2==0`), you might actually extend mention the parameter, and I think the hyphen confuses the error:

```
DG_CHECK(osnArgs.extraFields.size() % 2 == 0, "Extra fields must be supplied in pairs of key and value strings.");
```

File: common/src/OpenSpaceNet.cpp [hunk 2]
```diff
@@ -543,6 +543,23 @@ PredictionToFeature::Ptr OpenSpaceNet::initPredictionToFeature()
...
+        fields.emplace("app_ver", Field(FieldType::STRING, OPENSPACENET_VERSION_STRING));
+    }
+
+    if (args_.extraFields.size() > 0) {
+        for(int i = 0; i < args_.extraFields.size(); i += 2) {
+            std::string fieldKey = args_.extraFields[i];
+            std::string fieldVal = args_.extraFields[i + 1];
+            Field field(FieldType::STRING, fieldVal);
+            fields.emplace(fieldKey, field);
+
```
Comment (rddesmond) [inline common/src/OpenSpaceNet.cpp]: Extra line break

File: common/src/OpenSpaceNet.cpp [hunk 3]
```diff
@@ -543,6 +543,23 @@ PredictionToFeature::Ptr OpenSpaceNet::initPredictionToFeature()
...
     predictionToFeature->attr("topNCategories") = 5;
 
+    auto fields = predictionToFeature->attr("extraFields").cast<std::map<std::string, Field>>();
+    if(args_.producerInfo) {
+        fields.emplace("username", Field(FieldType::STRING, loginUser()));
+        fields.emplace("app", Field(FieldType::STRING, "OpenSpaceNet"));
+        fields.emplace("app_ver", Field(FieldType::STRING, OPENSPACENET_VERSION_STRING));
+    }
+
+    if (args_.extraFields.size() > 0) {
```
Comment (rddesmond) [inline common/src/OpenSpaceNet.cpp]: Clion says to me: use `!empty`.

File: common/src/OpenSpaceNet.cpp [hunk 4]
```diff
@@ -543,6 +543,23 @@ PredictionToFeature::Ptr OpenSpaceNet::initPredictionToFeature()
...
+        for(int i = 0; i < args_.extraFields.size(); i += 2) {
+            std::string fieldKey = args_.extraFields[i];
+            std::string fieldVal = args_.extraFields[i + 1];
+            Field field(FieldType::STRING, fieldVal);
```
Comment (rddesmond) [inline common/src/OpenSpaceNet.cpp]: Move `fieldVal`, `field`, `fieldKey` (or don't create the intermediate)

File: common/src/OpenSpaceNet.cpp [hunk 5]
```diff
@@ -543,6 +543,23 @@ PredictionToFeature::Ptr OpenSpaceNet::initPredictionToFeature()
...
+    auto fields = predictionToFeature->attr("extraFields").cast<std::map<std::string, Field>>();
+    if(args_.producerInfo) {
+        fields.emplace("username", Field(FieldType::STRING, loginUser()));
+        fields.emplace("app", Field(FieldType::STRING, "OpenSpaceNet"));
+        fields.emplace("app_ver", Field(FieldType::STRING, OPENSPACENET_VERSION_STRING));
+    }
+
+    if (args_.extraFields.size() > 0) {
+        for(int i = 0; i < args_.extraFields.size(); i += 2) {
+            std::string fieldKey = args_.extraFields[i];
```
Comment (rddesmond) [inline common/src/OpenSpaceNet.cpp]: C++11 `auto`

File: common/src/OpenSpaceNet.cpp [hunk 6]
```diff
@@ -623,6 +640,13 @@ FileFeatureSink::Ptr OpenSpaceNet::initFeatureSink()
         definitions.emplace_back(FieldType::STRING, "catalog_id");
     }
 
+    for (int i = 0 ; i < args_.extraFields.size(); i+=2){
+        std::string fieldKey = args_.extraFields[i];
```
Comment (rddesmond) [inline common/src/OpenSpaceNet.cpp]: Same comments as above wrt `move` and `auto`.

File: cli/src/CliProcessor.cpp (hunk 1 above)
Comment (geovations) [reply cli/src/CliProcessor.cpp]: I don't want to do anything if extraFields is empty, but I guess it's ok.

File: common/src/OpenSpaceNet.cpp (hunk 3 above)
Comment (geovations) [reply common/src/OpenSpaceNet.cpp]: That works!

File: common/src/OpenSpaceNet.cpp (hunk 6 above)
Comment (geovations) [reply common/src/OpenSpaceNet.cpp]: RGR

**Pull request merged**